crypto/init.c

```diff
@@ -78,13 +78,144 @@ DEFINE_RUN_ONCE_STATIC(ossl_init_base)
     return 0;
 }
 
+/*
+ * Store of provider handles for which there OPENSSL_cleanup() is set up as
+ * and exit function.  The handle NULL represents the application.
+ *
+ * This allows several components (application and providers alike) to share
+ * the same instance of libcrypto, and to ensure that OPENSSL_cleanup() isn't
+ * called too early.
+ */
+static const OSSL_CORE_HANDLE **handles = NULL;
+static size_t handles_num = 0;
+
+static int ossl_init_register_cleanup(const OSSL_CORE_HANDLE *handle)
+{
+    size_t i;
+    int ret;
+
+    if (!CRYPTO_THREAD_write_lock(init_lock))
+        return -1;
+
+#ifdef OPENSSL_INIT_DEBUG
+    fprintf(stderr,
+            "OPENSSL_INIT: ossl_init_register_cleanup(handle = %p)\n",
+            (void *)handle);
+#endif
+
+    /*
+     * Check if this handle is already registered, as we only allow
+     * one entry each...  Finding one is success.
+     */
+    ret = (int)handles_num;
+    for (i = 0; i < handles_num; i++)
+        if (handles[i] == handle)
+            break;
+
+    if (i < handles_num)
+        goto end;
+
+    /*
+     * Allocate the handles array in chunks of 10 elements.
+     * Trust that the memory allocator doesn't actually reallocate
+     * when the size is unchanged.
+     */
+    handles
```

Comment:
Why can't you use a stack for this?
It seems like you should be using one.

Reply:
Lots of extra churn.  Also, slight paranoia...  the init code is quite sensitive, I didn't want to add too much dependencies on the rest of libcrypto, on the off chance that we make changes in, for example, the stack code that wouldn't sit well here.

Reply:
I still think it would be better to use the well tested stack code rather than hacking something together here -- the comment about trusting the memory allocator indicates the level of hack.  A realloc implementation that always allocates a new block and uses memcpy to duplicate the final size is complaint with the standard.

If you want to avoid the stack code, at least use a singly linked list and allocate each node separately.  What's here is just bad.

Reply:
Actually, that comment is unnecessary, it's not really a crucial thing here.

A singly linked list seems like a lot of churn for a very simple array of addresses.

I'll start with removing that comment.

Reply:
The realloc can fail and this will leak the original handles allocation.

Reply:
Just use a stack, don't re-implement things unnecessarily.

Reply:
Either way, a decision needs to be made about what to do when allocation fails.  Currently, it happens "naturally", by crashing of `handles` ends up being NULL.  Whatever we do, if allocation fails, I can't see how to continue, as this array is essentially corrupt.  Changing this to a stack will make no difference in that case, except having to add some sort of explicit crash code.

Reply:
The stack code calls ERR_raise in some places...  that's the actual call that got us into this journey.  Calling something that leads there here has all my internal warning claxons go off at the same time.

So no, will not use the stack.

Reply:
Have a static array with, say, 100 elements.  No malloc/realloc no possible failure except overflow.

Reply:
Okie, done